hadoop-yarn-project/hadoop-yarn/hadoop-yarn-client/src/test/java/org/apache/hadoop/yarn/client/api/impl/TestNMClient.java

```diff
@@ -125,576 +117,344 @@ public void postTransition(
-    
-    //propagating the AMRMClient NMTokenCache instance
     nmClient.setNMTokenCache(rmClient.getNMTokenCache());
     nmClient.init(conf);
     nmClient.start();
-    assertNotNull(nmClient);
     assertEquals(STATE.STARTED, nmClient.getServiceState());
   }
 
-  @After
-  public void tearDown() {
+  public void tearDown() throws InterruptedException {
     rmClient.stop();
     yarnClient.stop();
-    yarnCluster.stop();
-  }
-
-  private void stopNmClient(boolean stopContainers) {
-    assertNotNull("Null nmClient", nmClient);
-    // leave one unclosed
-    assertEquals(1, nmClient.startedContainers.size());
-    // default true
-    assertTrue(nmClient.getCleanupRunningContainers().get());
-    nmClient.cleanupRunningContainersOnStop(stopContainers);
-    assertEquals(stopContainers, nmClient.getCleanupRunningContainers().get());
-    nmClient.stop();
+    yarnCluster.asyncStop(this);
   }
 
-  @Test (timeout = 180000)
+  @Test (timeout = 180_000 * MAX_EARLY_FINISH)
   public void testNMClientNoCleanupOnStop()
-      throws YarnException, IOException {
-
-    rmClient.registerApplicationMaster("Host", 10000, "");
+      throws YarnException, IOException, InterruptedException, TimeoutException {
+    runTest(() -> {
+      stopNmClient();
+      assertFalse(nmClient.startedContainers.isEmpty());
+      nmClient.cleanupRunningContainers();
+      assertEquals(0, nmClient.startedContainers.size());
+    });
+  }
 
-    testContainerManagement(nmClient, allocateContainers(rmClient, 5));
+  @Test (timeout = 200_000 * MAX_EARLY_FINISH)
+  public void testNMClient()
+      throws YarnException, IOException, InterruptedException, TimeoutException {
+    runTest(() -> {
+      // stop the running containers on close
+      assertFalse(nmClient.startedContainers.isEmpty());
+      nmClient.cleanupRunningContainersOnStop(true);
+      assertTrue(nmClient.getCleanupRunningContainers().get());
+      nmClient.stop();
+    });
+  }
 
-    rmClient.unregisterApplicationMaster(FinalApplicationStatus.SUCCEEDED,
-        null, null);
-    // don't stop the running containers
-    stopNmClient(false);
-    assertFalse(nmClient.startedContainers.isEmpty());
-    //now cleanup
-    nmClient.cleanupRunningContainers();
-    assertEquals(0, nmClient.startedContainers.size());
+  public void runTest(
+      Runnable test
+  ) throws IOException, InterruptedException, YarnException, TimeoutException {
+    int earlyFinishCounter = MAX_EARLY_FINISH;
+    int earlyFinishCounterWhenTestWasStarted;
+    do {
+      earlyFinishCounterWhenTestWasStarted = earlyFinishCounter;
+      setup();
+      rmClient.registerApplicationMaster("Host", 10_000, "");
+      testContainerManagement(nmClient, allocateContainers(rmClient));
+      rmClient.unregisterApplicationMaster(FinalApplicationStatus.SUCCEEDED, null, null);
+      test.run();
+      tearDown();
+    } while (earlyFinishCounter != 0 && earlyFinishCounter != earlyFinishCounterWhenTestWasStarted);
```

Comment:
Condition 'earlyFinishCounter != 0' is always 'true' 
Condition 'earlyFinishCounter != earlyFinishCounterWhenTestWasStarted' is always 'false' when reached

Reply:
Looks like this earlyFinishCounter was added then removed.

Here is a patch with some of my review comment fixes: https://github.com/tomicooler/hadoop/commit/cc49a495178dbc25e7a896dc2c2f5f9b3ad67e1c

Reply:
Cherry-picked this commit and removed few lines of System.out.println. Thanks for this commit.